// ngx-ui-loader/esm5/lib/core/ngx-ui-loader.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Task() { }
/**
 * @record
 */
export function Loader() { }
if (false) {
    /** @type {?} */
    Loader.prototype.loaderId;
    /** @type {?} */
    Loader.prototype.foreground;
    /** @type {?} */
    Loader.prototype.background;
    /** @type {?|undefined} */
    Loader.prototype.isMaster;
    /** @type {?} */
    Loader.prototype.isBound;
}
/**
 * @record
 */
export function Loaders() { }
/**
 * @record
 */
function LoaderEvent() { }
if (false) {
    /** @type {?} */
    LoaderEvent.prototype.loaderId;
}
/**
 * @record
 */
export function ShowEvent() { }
if (false) {
    /** @type {?} */
    ShowEvent.prototype.isShow;
}
/**
 * @record
 */
export function StartStopEvent() { }
if (false) {
    /** @type {?} */
    StartStopEvent.prototype.taskId;
    /** @type {?} */
    StartStopEvent.prototype.isForeground;
}
